README.md

```diff
@@ -93,75 +105,106 @@ internal GitHub server errors), the bot sleeps for 10 minutes
 configuration is preserved across such restarts.
 
 If a PR processing step fails for PR-specific reasons (e.g., a CI test
-failure), then the bot moves on to the next pull request, marking the
+failure), then the bot moves on to the next pull request, labeling the
 failed PR if/as needed (see below for PR labels).
 
 
 ## Pull request labels
 
-The bot uses the following GitHub labels to mark the current pull
+The bot uses the following custom GitHub labels to mark the current pull
 request state:
 
-* S-merging: The PR is being processed by the bot. The bot eventually
-  replaces this label with one of the other labels below.
-* S-staging-checks-failed: Essentially duplicates GitHub "red x" mark for
-  the staging branch commit. The bot removes this label when it notices
+* `M-waiting-approvals`: The PR approval rules (see "Voting and PR
+  approvals" below) do not allow the bot to start merging regardless of
+  how old the PR gets. Compare with `M-waiting-objections`. The bot
+  removes this label when it notices that it is obsolete. TODO:
+  Implement.
+* `M-waiting-objections`: The PR approval rules (see "Voting and PR
+  approvals" below) allow the bot to start merging when the PR gets
+  older. The bot will eventually start merging this PR if no objections
+  are received while the bot is waiting. The wait time is determined by
+  the number of votes already received. The bot removes this label when
+  it notices that it is obsolete. TODO: Implement.
+* `M-waiting-staging-checks`: The bot determined that the PR is eligible
+  for merging, successfully completed the initial merging steps that are
+  supposed to trigger CI tests, and is now waiting for the CI test
+  results. The bot removes this label when it notices that all tests
+  have completed.
+* `M-passed-staging-checks`: Similar to the GitHub "green check" mark
+  for the staging branch commit (but ignores failures of optional
+  checks). Applied only if `config::skip_merge` option is on. The bot
+  removes this label when applying any other label (XXX: Except
+  M-failed-staging-checks. To improve code consistency and to avoid
+  lying in documentation, we should either remove this label only when
+  (re)setting the first label (i.e., M-waiting-staging-checks) or remove
+  this label when setting any other label, without making an exception
+  for M-failed-staging-checks. The implementation plans for the
```

Comment:
I would not do as you suggested, because:
* "removing the label only when resetting the first label" misses a case when we should remove it after PR is merged(and closed): we should not such excessive details after completion
* "removing when setting any other" (including M-failed-staging-checks) also looks excessive because I do not see how the passed checks suddenly become 'failed'.

Instead I would suggest to fix existing code and the documentation: IMO, we need to remove this label only in two cases:
* M-merged
* M-failed-other
because these should be the only possible states for PR after it's staging checks have succeeded.

Reply:
> "removing the label only when resetting the first label" misses a case when we should remove it after PR is merged

I would expect the first label (i.e., `M-waiting-staging-checks`) to be (re)set when 

* the bot (re)starts waiting for those checks (because something has changed and new tests are needed) or 
* the bot decides to proceed beyond staging checks (because the `config::skip_merge` option was removed). Since the bot does not read/check `M-waiting-staging-checks`, it has to reset it when proceeding beyond the staging checks, right?

Both events must happen way before the PR is merged. Am I missing something?

> "removing when setting any other" also looks excessive 

It certainly is excessive! However, it is also very simple. If removing a non-existent label does not generate a user-visible notification on GitHub, then it is a viable option because of its simplicity.

> IMO, we need to remove this label only in two cases: `M-merged` and `M-failed-other`

... and `M-waiting-status-checks` and `M-failed-description` (at least).

If you want to minimize `M-passed-staging-checks` removals, then we should remove this label when it is set but is no longer valid. It is no longer valid when the _tested_ commit is no longer fresh/applicable. I do not recommend this path because it probably increases code complexity.